### ISSUE TITLE
Fuji: bit fibermask.BADPOSITION not set for some fibers/exposures with missing positioning data

### ISSUE THREAD

**Issue by julienguy** [desihub/desispec#1676]
The FIBERSTATUS bit fibermask.BADPOSITION is not set for some fibers/exposures with missing positioning data.
Example:

```
from astropy.table import Table
fmap=Table.read(f"/global/cfs/cdirs/desi/spectro/redux/fuji/tiles/cumulative/80605/20210205/spectra-7-80605-thru20210205.fits","FIBERMAP")
tid=39627676691992850
print(fmap["FIBERSTATUS"][fmap["TARGETID"]==tid])
print(fmap["FIBER_X"][fmap["TARGETID"]==tid])
```

```
  0
  0
  0
  0
  0
  0
  0
  0
  0
  0
  0
  0
512
512
512
512
512
512
512
512
512
512
512
512

fmap["FIBER_X"][fmap["TARGETID"]==tid]
     0.0
     0.0
     0.0
     0.0
     0.0
     0.0
     0.0
     0.0
     0.0
     0.0
     0.0
     0.0
-188.342
-188.355
-189.839
-189.836
 -189.83
-189.833
-189.839
-189.837
-189.843
 -189.85
-189.839
-189.832


```

**Comment by sbailey** [comment]
Argh.  Astropy 5.0 broke this through their Table "feature" that NaN entries are read as masked values instead of just keeping them as NaN.  Astropy 4.0.x used to just preserve them as NaN and we were checking for bad entries with `np.isnan`, but with astropy 5.0 that returns False instead of True because the NaNs are masked...

excerpts from desispec.io.fibermap.assemble_fibermap:
```
        pm = Table.read(coordfile, 'DATA')  #- PM = PlateMaker
        fibermap['DELTA_X'] = pm[f'DX_{numiter-1}']
        fibermap['DELTA_Y'] = pm[f'DY_{numiter-1}']
        dr = np.sqrt(fibermap['DELTA_X']**2 + fibermap['DELTA_Y']**2) * 1000
        badpos = (dr > bad_offset_um) | np.isnan(dr)
```

With astropy 4.0.x this identified input platemaker NaNs as badpos, but now fails to do that.

**Comment by julienguy** [comment]
I scanned all Fuji spectra files, there are 19 affected tiles are:
```
80605
80606
80607
80608
80609
80610
80611
80612
80613
80614
80615
80616
80617
80618
80619
80620
80621
80622
80623
```

**Comment by sbailey** [comment]
Just to drive home the point of how messed up the astropy Table behavior is:
```
In [1]: import fitsio

In [2]: from astropy.table import Table

In [3]: fm1 = fitsio.read('fibermap-00067710.fits', 'FIBERMAP')

In [4]: fm2 = Table.read('fibermap-00067710.fits', 'FIBERMAP')

In [5]: np.any(np.isnan(fm1['DELTA_X']))
Out[5]: True

In [6]: np.any(np.isnan(fm2['DELTA_X']))
Out[6]: False
```

**Comment by julienguy** [comment]
I looked in the spectra files for cases with 
```
((t["FIBER_X"]==0)&(t["FIBER_Y"]==0)&(t["OBJTYPE"]=="TGT")&(t["FIBERSTATUS"]&(fibermask.BADPOSITION|fibermask.POORPOSITION|fibermask.MISSINGPOSITION)==0))
```

**Comment by weaverba137** [comment]
If this is causing severe difficulty, as opposed to just being an unexpected API change (remember, it's a major version change!), please report it to the Astropy collaboration.

**Comment by julienguy** [comment]
I checked that there is no tile in guadalupe affected by the bug. it's all contained to 20 tiles of fuji (and the last one is just one fiber but we may as well rerun it or fix it with the others).

**Comment by sbailey** [comment]
Those tiles were observed on 286 exposures on 20 nights spanning 20201214 through 20210331.  They are in cmx/other, sv1/bright, and sv1/dark , touching 208 healpix (the tiles, not necessarily the bad fibers).  Code snippet to identify healpix:
```
badtiles = [80605,80606,80607,80608,80609,80610,80611,80612,80613,80614,80615,80616,80617,80618,80619,80620,80621,80622,80623]
 
tilepixfiles = sorted(glob.glob('preproc/*/*/tilepix-*.json'))
healpix = set()
for filename in tilepixfiles:
    tileid = int(filename.split('-')[-1].split('.')[0])
    if tileid in badtiles:
        with open(filename) as fx:
            tilepix = json.load(fx)
        for p in range(10):
            healpix.update( set(tilepix[str(tileid)][str(p)]) )
```

I suspect that this was a purposeful API change by astropy, but I find it a big backwards step in usefuless.  We previously got bitten by this by astropy Tables assuming blank strings should be masked values instead of just actual blank strings, but fortunately we caught that one in time.

**Comment by sbailey** [comment]
Thinking aloud about options for Fuji
  1. live with it as a bug discovered too late
  2. intercept the NaNs at the time of creating the spectra file, and set the missing fiberstatus bits so that we only need to rerun redshifts
      * desispec.pixgroup already intercepts NaNs and sets them to 0.0; it's just a matter of also setting the fiberstatus bit
      * this would fix the spectra, coadds, and redrock but not upstream cframes etc.
      * it would also leave poorly positioned sky and stdstars are potentially being used because they weren't masked (ugh)
  3. bit the bullet and reprocess those tiles and healpix with fixed code...

I'm leaning towards (3) but don't want to launch into that hastily...

**Comment by sbailey** [comment]
@rongpu please check whether any of these tiles are in your set of custom 1x_depth and 4x_depth coadds.  If so, you'll need to re-run after we reprocess them.

**Comment by sbailey** [comment]
@rongpu `Table(fitsio.read(fn))` works great if you only need the *data* part of the table.  `desispec.io.table.read_table(fn)` basically does that, but then also propagates the header keywords into `table.meta` like you get with `Table.read(fn)` but you don't get with `Table(fitsio.read(fn))`.

fibermask.BADPOSITION fixed in PR #1681; closing.
